### PR TITLE
GIX-1116: Syns SNS Neuron Balances

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

Check sns neuron balances with the stake and claim or refresh accordingly if they don't match

# Changes

* New sns-governance api: "getNeuronBalance", "refreshNeuron", "claimNeuron".
* Rename sns neuron service "loadSnsNeurons" to "syncSnsNeurons". Add balance check there.
* Add balance check in getSnsNeuron service.
* Balance check and refreshing is done in "checkSnsNeuronBalances". No recursion, each neuron refreshed is loaded independently.
* New method in sns neurons store "addNeurons".

# Tests

* Test new cases in syncSnsNeurons and getSnsNeuron services.
* Test new sns neurons method.
* Test new sns governance apis.

### CODE REVIEW

**Comment by lmuntaner** [comment]
@mstrasinskis @peterpeterparker please review

PENDING: Tests, but the logic is already working.

**Comment by lmuntaner** [comment]
@mstrasinskis ready for full review

File: frontend/src/lib/services/sns-neurons.services.ts [hunk 1]
```diff
@@ -70,6 +86,226 @@ export const loadSnsNeurons = async (
...
+        rootCanisterId,
+        neuronId: currentNeuronId,
+        certified: false,
+        identity,
+      });
+      if (currentBalance > BigInt(0)) {
+        const neuron = neurons.find(
+          (neuron) =>
+            getSnsNeuronIdAsHexString(neuron) ===
+            subaccountToHexString(subaccount)
```
Comment (mstrasinskis) [inline frontend/src/lib/services/sns-neurons.services.ts:246]: Just a suggestion and probably not part of this PR. Because of it's own scope, what do you think if all the sns neuron-stake related entities are prefixed with `ns` (e.g. `nsSubaccountToHexString`)?

File: frontend/src/lib/services/sns-neurons.services.ts [hunk 2]
```diff
@@ -10,29 +13,35 @@ import {
...
   type SnsNeuronId,
 } from "@dfinity/sns";
-import { arrayOfNumberToUint8Array } from "@dfinity/utils";
+import { arrayOfNumberToUint8Array, fromNullable } from "@dfinity/utils";
 import { get } from "svelte/store";
 import { getIdentity } from "./auth.services";
 import { queryAndUpdate } from "./utils.services";
 
-export const loadSnsNeurons = async (
+export const syncSnsNeurons = async (
```
Comment (mstrasinskis) [inline frontend/src/lib/services/sns-neurons.services.ts:44]: Probably worth to add a comment that calling this function is the only way to reliable finalise transactions and get the actual neuron stakes. Just to emphasise its importance.
(Ideally also a short algorithm description from the doc because its implementation is not limited to one function)

File: frontend/src/lib/services/sns-neurons.services.ts [hunk 3]
```diff
@@ -70,6 +86,226 @@ export const loadSnsNeurons = async (
...
+  const visitedSubaccounts = new Set<string>();
+  // TODO: Check neurons controlled by linked HW?
+  const identity = await getIdentity();
+  const controller = identity.getPrincipal();
+  // Visit subaccounts by order until we find a balance of 0.
+  // Therefore, we set the initial balance to some non-zero value.
+  let currentBalance = BigInt(1);
+  for (
+    let index = 0;
+    index < MAX_NEURONS_SUBACCOUNTS && currentBalance > BigInt(0);
```
Comment (mstrasinskis) [inline frontend/src/lib/services/sns-neurons.services.ts:225]: Probably using `break` instead of `currentBalance > BigInt(0)` could simplify the code a bit.

File: frontend/src/lib/services/sns-neurons.services.ts [hunk 4]
```diff
@@ -70,6 +86,226 @@ export const loadSnsNeurons = async (
...
+      visitedSubaccounts.add(subaccountToHexString(subaccount));
+      // Id of an sns neuron is the subaccount of the sns governance canister where the stake is
+      const currentNeuronId: SnsNeuronId = { id: subaccount };
+      currentBalance = await getNeuronBalance({
+        rootCanisterId,
+        neuronId: currentNeuronId,
+        certified: false,
+        identity,
+      });
+      if (currentBalance > BigInt(0)) {
```
Comment (mstrasinskis) [inline frontend/src/lib/services/sns-neurons.services.ts:242]: nitpick: would be nice to to have more consts for readability. e.g.
```
const positiveBalance: boolean
const knownNeuron: boolean
const differentBalance: boolean
```

File: frontend/src/lib/services/sns-neurons.services.ts [hunk 5]
```diff
@@ -70,6 +86,226 @@ export const loadSnsNeurons = async (
...
+            controller,
+            memo: BigInt(index),
+            subaccount,
+          });
+        } else {
+          // Check if neuron needs refresh.
+          const neuronId = fromNullable(neuron.id);
+          if (neuronId !== undefined) {
+            if (
+              await neuronNeedsRefresh({
```
Comment (mstrasinskis) [inline frontend/src/lib/services/sns-neurons.services.ts:262]: Wouldn't it check the same balance as loaded before #236?
https://github.com/dfinity/nns-dapp/pull/1529/files#diff-3a1a9c5c924b5918e75d65e6497c777fa172c89b58d7a649e8ffa9b5c97f9359R236

File: frontend/src/lib/services/sns-neurons.services.ts [hunk 6]
```diff
@@ -70,6 +86,226 @@ export const loadSnsNeurons = async (
...
+  });
+  snsNeuronsStore.addNeurons({
+    rootCanisterId,
+    certified,
+    neurons: [neuron],
+  });
+};
+
+/**
+ * Returns true if the neuton's stake doesn't match the balance of the subaccount.
```
Comment (mstrasinskis) [inline frontend/src/lib/services/sns-neurons.services.ts:133]:
```suggestion
 * Returns true if the neuron's stake doesn't match the balance of the subaccount.
```

File: frontend/src/lib/services/sns-neurons.services.ts [hunk 7]
```diff
@@ -70,6 +86,226 @@ export const loadSnsNeurons = async (
...
+          }
+        }
+      }
+    } catch (error) {
+      // Ignore the error, we do this in the background.
+      // If this fails the data might be stale but it should refreshed on the next sync.
+      console.error(error);
+    }
+  }
+  // Check unvisited neurons.
```
Comment (mstrasinskis) [inline frontend/src/lib/services/sns-neurons.services.ts:285]: nitpick: Probably worth a separate function. This one is already pretty big.

File: frontend/src/lib/services/sns-neurons.services.ts [hunk 8]
```diff
@@ -70,6 +86,226 @@ export const loadSnsNeurons = async (
...
+            }
+          }
+        }
+      }
+    } catch (error) {
+      // Ignore the error, we do this in the background.
+      // If this fails the data might be stale but it should refreshed on the next sync.
+      console.error(error);
+    }
+  }
```
Comment (mstrasinskis) [inline frontend/src/lib/services/sns-neurons.services.ts:284]: Is this out of the current PR scope?
> if subaccount in neurons:
>      neurons.remove(subaccount)

File: frontend/src/lib/services/sns-neurons.services.ts (hunk 2 above)
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons.services.ts:44]: That's a good idea, I'll add it.

File: frontend/src/lib/services/sns-neurons.services.ts [hunk 9]
```diff
@@ -70,6 +86,201 @@ export const loadSnsNeurons = async (
   });
 };
 
+const loadNeuron = async ({
+  rootCanisterId,
```
[the remark this answers is not in the document]
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons.services.ts:110]: Sounds good. Initially I thought I might export it and use it somewhere else. That's why I didn't add it. But for now it's more of a helper function.

File: frontend/src/lib/services/sns-neurons.services.ts (hunk 6 above)
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons.services.ts:133]: 😂

File: frontend/src/lib/services/sns-neurons.services.ts [hunk 10]
```diff
@@ -70,6 +86,201 @@ export const loadSnsNeurons = async (
...
+      (await getNeuronBalance({
+        rootCanisterId,
+        neuronId,
+        certified: false,
+        identity,
+      }));
+    return balance !== neuron.cached_neuron_stake_e8s;
+  }
+  // Edge case, a valid neuron should have a neuron id
+  return false;
```
[the remark this answers is not in the document]
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons.services.ts:167]: 👍

File: frontend/src/lib/services/sns-neurons.services.ts [hunk 11]
```diff
@@ -70,6 +86,226 @@ export const loadSnsNeurons = async (
...
+  rootCanisterId: Principal;
+  neurons: SnsNeuron[];
+}): Promise<void> => {
+  const visitedSubaccounts = new Set<string>();
+  // TODO: Check neurons controlled by linked HW?
+  const identity = await getIdentity();
+  const controller = identity.getPrincipal();
+  // Visit subaccounts by order until we find a balance of 0.
+  // Therefore, we set the initial balance to some non-zero value.
+  let currentBalance = BigInt(1);
```
[the remark this answers is not in the document]
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons.services.ts:222]: Sounds good

File: frontend/src/lib/services/sns-neurons.services.ts (hunk 3 above)
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons.services.ts:225]: I had it, but it obscured a little bit the condition to break the loop.

I found the logic clearer like this. A break seems like an exception, but in this loop, that's exactly when we want to stop, more than the `MAX_NEURONS_SUBACCOUNTS`.

I almost used a while loop, but we also needed the index to go up, that's why I used the for loop.

File: frontend/src/lib/services/sns-neurons.services.ts [hunk 12]
```diff
@@ -70,6 +86,201 @@ export const loadSnsNeurons = async (
...
+  for (
+    let index = 0;
+    index < MAX_NEURONS_SUBACCOUNTS && currentBalance > BigInt(0);
+    index++
+  ) {
+    try {
```
[the remark this answers is not in the document]
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons.services.ts:228]: Good point

File: frontend/src/lib/services/sns-neurons.services.ts (hunk 4 above)
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons.services.ts:242]: All in for readability, sounds good.

File: frontend/src/lib/services/sns-neurons.services.ts (hunk 1 above)
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons.services.ts:246]: `subaccountToHexString` has nothing to do with neurons.

We could use this for SNS subaccounts as well.

Do you mean to diffentiate utils that are for SNS and for NNS?

File: frontend/src/lib/services/sns-neurons.services.ts (hunk 5 above)
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons.services.ts:262]: Yes it does.

That's why I pass it as parameter:

```
const needsRefresh = await neuronNeedsRefresh({
  rootCanisterId,
  neuron,
  identity,
  // Here I pass the balance
  balanceE8s: currentBalance,
});
```

I used a helper to reuse the logic of what it means to need a refresh.

But maybe I could use an util that expects the balance, and use it here and use the service that also fetches the balance in the other places.

File: frontend/src/lib/services/sns-neurons.services.ts (hunk 8 above)
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons.services.ts:284]: This probably depends on the error.

But because this is an extra check, I think that if chekcing balances fail, we are safe to assume that the update call to get the neurons is a good fallback.

File: frontend/src/lib/services/sns-neurons.services.ts (hunk 7 above)
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons.services.ts:285]: Sounds good, let me refactor a bit.

**Comment by lmuntaner** [comment]
@mstrasinskis please review. I did quite a bit of refactor to try to better explain what's going on. Let me know what you think.

File: frontend/src/lib/services/sns-neurons.services.ts (hunk 3 above)
Comment (mstrasinskis) [reply frontend/src/lib/services/sns-neurons.services.ts:237]: Thanks for the explanation

File: frontend/src/lib/services/sns-neurons.services.ts (hunk 1 above)
Comment (mstrasinskis) [reply frontend/src/lib/services/sns-neurons.services.ts:261]: I thought that it is a bit confusing because `subaccountToHexString` also has nothing to do with account subaccounts. But not a strong opinion.

Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons.services.ts]: discussed in a call

File: frontend/src/lib/services/sns-neurons.services.ts (hunk 8 above)
Comment (lmuntaner) [reply frontend/src/lib/services/sns-neurons.services.ts]: Done. The algorithm refreshes the neuron and if the stake if 0 it will not show in the UI.

**Comment by lmuntaner** [comment]
@mstrasinskis please review

I have refactored the code to its own file.

**Review by mstrasinskis — APPROVED** [review]
LGTM. Special thanks for the documentation update!
